cpp/serve/sampler/gpu_sampler.cc

```diff
@@ -153,11 +223,13 @@ class GPUSampler : public SamplerObj {
     // - Check if there is need for applying top p or prob values,
     //   so that argsort is needed.
     bool need_top_p = false;
-    bool need_prob_values = false;
+    if (!top_p_applied) {
+      need_top_p = CheckTopP(generation_cfg, sample_indices, num_probs, num_samples, vocab_size);
+    }
     // The indptr array of the number of top probs for each sample.
     std::vector<int> top_prob_offset_indptr;
-    CheckTopPAndProbValues(generation_cfg, sample_indices, num_probs, num_samples, vocab_size,
-                           &need_top_p, &need_prob_values, &top_prob_offset_indptr);
+    bool need_prob_values = CheckProbValues(generation_cfg, sample_indices, num_probs, num_samples,
```

Comment:
Note (for future refactor): this will cause argsort be applied again to get the top probs

Reply:
Yes. For now we don't have a good choice besides sort.